<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-programming/genomic_data/Getting_started_with_genomic_data">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">Getting started with genomic data | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="Getting started with genomic data | WHG training resources"><meta data-rh="true" name="description" content="Authors: Ben Wright, Wojciech Lason Bioinformatics Core"><meta data-rh="true" property="og:description" content="Authors: Ben Wright, Wojciech Lason Bioinformatics Core"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.4b0609aa.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.a59577cb.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/">programming</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/">genomic_data</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/">Getting started with genomic data</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/">Introduction to R</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to R&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/">Programming with gene annotations</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming with gene annotations&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">programming</span><meta itemprop="position" content="1"></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">genomic_data</span><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">Getting started with genomic data</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>Getting started with genomic data</h1><p>Authors: Ben Wright, Wojciech Lason Bioinformatics Core</p><h1>Session aims</h1><ul><li>Introduce common file formats and their uses</li><li>Introduce command line tools to work with those files</li><li>Put bash and R skills to use filtering variant data</li></ul><p>The intention is not to teach the specifics of every format and every
tool, but to make you<br>
<!-- -->familiar with the landscape of bioinfoamtics tools and data so if you
need to solve a specific problem in the course of research you have an
idea of where to start.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="text-formats">Text formats<a class="hash-link" href="#text-formats" title="Direct link to heading">​</a></h2><p>Many data files used in genomics are in &#x27;plain text&#x27; format. They are
ordinary text with spaces or tabs used to separate columns. The latter
is often referred to as <code>.tsv</code> format, although it&#x27;s seldom encountered
with that actual file suffix. The advantage to plain text is that you
can read it on the command line or load it into a spreadsheet programme
easily. The downside is that this is not generally an efficient storage
format and leads to large file sizes. It&#x27;s not at all uncommon to work
with text files so large that common text editors struggle to open them,
even on modern computers.</p><p>While it&#x27;s entirely possible to manipulate these files using the
standard command line tools like <code>cut</code> or <code>awk</code>, you probably don&#x27;t
actually want to do that if you can help it. There are better tools
available, designed to work with specific types of file.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="binary-formats">Binary formats<a class="hash-link" href="#binary-formats" title="Direct link to heading">​</a></h2><p>Many text formats have an equivalent &#x27;binary&#x27; format. This holds the
same data, but not in plain text you can read in a terminal or editor.
Binary formats are more efficient storage, but require specific tools to
read and work with. These tools always have a function to convert back
into plain text format so you can inspect the data.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="compression">Compression<a class="hash-link" href="#compression" title="Direct link to heading">​</a></h2><p>For large files, it&#x27;s common to work with a compressed version of the
file. This doesn&#x27;t just help reduce the space needed to store the files.
On modern computers, reading a compressed file from disk and
decompressing it is faster than reading the same uncompressed data from
disk (almost always). On the command line, you usually can&#x27;t work with
compressed files directly, even compressed text files. You need to
decompress the file using a suitable tool then use pipes (<code>|</code>) to the
command you wanted. If you want to save a compressed version of a file,
you need to pipe the output to a command that compresses the data before
sending it to a file.</p><p>You have probably encountered the common &#x27;.zip&#x27; compression format
before. This format, and other compressed formats like it, have a
drawback when it comes to working with genomic data: you have to
decompress the entire file even when you only want a small part of it.
When genomic data is compressed, it is often done using a &#x27;block
compression&#x27; algorithm. What block compression does is slice the data up
into blocks and compress each block separately, before saving them in
the same file. Therefore, you only need to decompress the block that
holds the data you are interested in. This makes accessing random parts
of a large, compressed file much more efficient.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="index-files">Index files<a class="hash-link" href="#index-files" title="Direct link to heading">​</a></h3><p>Command line tools still need information about where they need to look
in block compressed formats. It&#x27;s very common for compressed genomic
data to be accompanied by a separate &#x27;index&#x27; file. This index file is
generated from the main compressed file, and just contains information
on what is stored in each block. It&#x27;s this index that lets a tool know
where in the block compressed file it needs to look. In genomic data,
which usually has some positional information in terms of chromosomal
position, this is very convenient.</p><p>You commonly need to run a separate command to create an index after you
have created a compressed file.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="file-headers">File headers<a class="hash-link" href="#file-headers" title="Direct link to heading">​</a></h2><p>Depending on the format, these files may or may not have a header row
that names the columns. Where they do not, this is because the columns,
and their order, are defined precisely in the format specification.</p><p>It is also common for these files to have a header section before the
table itself. This is used to store meta-data about the rest of the
file. This <em>could</em> be kept in a separate file, but that makes it too
easy for the two files to become separated. Many tools use information
in this header to make sure they are interpreting the data correctly.</p><p>As these headers are at the start of the file, it is simple to view them
from the command line, allowing you to check this metadata for yourself
to make sure it is what&#x27;s expected.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="versioning">Versioning<a class="hash-link" href="#versioning" title="Direct link to heading">​</a></h2><p>Many data files will be created with reference to a specific build of
the reference genome for an organism. It&#x27;s vital to avoid mixing up
these versions between different steps of data processing. Even if the
organism is the same, and the sequence very similar, the positions of
features such as genes can vary considerably and this can be a source of
errors.</p><h1>Common genomics file formats</h1><p>It&#x27;s not expected that you have all the details of file formats
committed to memory. It <em>is</em> useful to be able to recognise most common
ones and know where to look to find the detailed information you might
need.</p><p>Fuller descriptions of these formats can be found
<a href="https://learn.gencore.bio.nyu.edu/ngs-file-formats/" target="_blank" rel="noopener noreferrer">here</a>.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="bed-format">BED format<a class="hash-link" href="#bed-format" title="Direct link to heading">​</a></h2><p><a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1" target="_blank" rel="noopener noreferrer">BED format
definition</a></p><ul><li>Suffix: <code>.bed</code></li><li>Header: none</li><li>Uses: defining regions of interest (such as targetted sequencing
kits), output from feature detection algorithms</li><li>Acquired from: downloads of sequencing kit targets, databases of
genomic features</li></ul><p>This format defines genomic regions as simply as possible. Chromosome
identifer (or equivalent), start position and end position. It can
optionally have extra columns with additional information about those
regions, such as strand or defining sub-features.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="bed-example">BED example<a class="hash-link" href="#bed-example" title="Direct link to heading">​</a></h3><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127471196  127472363  Pos1  0  +  127471196  127472363  255,0,0</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127472363  127473530  Pos2  0  +  127472363  127473530  255,0,0</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127473530  127474697  Pos3  0  +  127473530  127474697  255,0,0</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127474697  127475864  Pos4  0  +  127474697  127475864  255,0,0</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127475864  127477031  Neg1  0  -  127475864  127477031  0,0,255</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127477031  127478198  Neg2  0  -  127477031  127478198  0,0,255</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127478198  127479365  Neg3  0  -  127478198  127479365  0,0,255</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127479365  127480532  Pos5  0  +  127479365  127480532  255,0,0</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">chr7    127480532  127481699  Neg4  0  -  127480532  127481699  0,0,255</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="gff3-format">GFF3 format<a class="hash-link" href="#gff3-format" title="Direct link to heading">​</a></h2><p><a href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md" target="_blank" rel="noopener noreferrer">GFF3 format
definition</a></p><ul><li>Suffix: <code>.gff3</code>, <code>.gff</code> or <code>.gtf</code></li><li>Header: <code>##</code> comments providing context for the file such as version
numbers and region identifiers</li><li>Uses: defining genomic features such as genes and exons</li><li>Acquired from: downloads of gene/exon/feature definitions</li></ul><p>This set of related formats also defines genomic regions, like <code>.bed</code>
format, but has additional required columns with information regarding
feature type, strand and phase.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="gff3-example">GFF3 example<a class="hash-link" href="#gff3-example" title="Direct link to heading">​</a></h3><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">##gff-version 3.1.26</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##sequence-region ctg123 1 1497228</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . gene            1000  9000  .  +  .  ID=gene00001;Name=EDEN</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . TF_binding_site 1000  1012  .  +  .  ID=tfbs00001;Parent=gene00001</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . mRNA            1050  9000  .  +  .  ID=mRNA00001;Parent=gene00001;Name=EDEN.1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . mRNA            1050  9000  .  +  .  ID=mRNA00002;Parent=gene00001;Name=EDEN.2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . mRNA            1300  9000  .  +  .  ID=mRNA00003;Parent=gene00001;Name=EDEN.3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . exon            1300  1500  .  +  .  ID=exon00001;Parent=mRNA00003</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . exon            1050  1500  .  +  .  ID=exon00002;Parent=mRNA00001,mRNA00002</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . exon            3000  3902  .  +  .  ID=exon00003;Parent=mRNA00001,mRNA00003</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . exon            5000  5500  .  +  .  ID=exon00004;Parent=mRNA00001,mRNA00002,mRNA00003</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . exon            7000  9000  .  +  .  ID=exon00005;Parent=mRNA00001,mRNA00002,mRNA00003</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             1201  1500  .  +  0  ID=cds00001;Parent=mRNA00001;Name=edenprotein.1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             3000  3902  .  +  0  ID=cds00001;Parent=mRNA00001;Name=edenprotein.1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             5000  5500  .  +  0  ID=cds00001;Parent=mRNA00001;Name=edenprotein.1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             7000  7600  .  +  0  ID=cds00001;Parent=mRNA00001;Name=edenprotein.1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             1201  1500  .  +  0  ID=cds00002;Parent=mRNA00002;Name=edenprotein.2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             5000  5500  .  +  0  ID=cds00002;Parent=mRNA00002;Name=edenprotein.2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             7000  7600  .  +  0  ID=cds00002;Parent=mRNA00002;Name=edenprotein.2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             3301  3902  .  +  0  ID=cds00003;Parent=mRNA00003;Name=edenprotein.3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             5000  5500  .  +  1  ID=cds00003;Parent=mRNA00003;Name=edenprotein.3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             7000  7600  .  +  1  ID=cds00003;Parent=mRNA00003;Name=edenprotein.3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             3391  3902  .  +  0  ID=cds00004;Parent=mRNA00003;Name=edenprotein.4</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             5000  5500  .  +  1  ID=cds00004;Parent=mRNA00003;Name=edenprotein.4</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ctg123 . CDS             7000  7600  .  +  1  ID=cds00004;Parent=mRNA00003;Name=edenprotein.4</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="fasta-format">FASTA format<a class="hash-link" href="#fasta-format" title="Direct link to heading">​</a></h2><p><a href="https://blast.ncbi.nlm.nih.gov/Blast.cgi?CMD=Web&amp;PAGE_TYPE=BlastDocs&amp;DOC_TYPE=BlastHelp" target="_blank" rel="noopener noreferrer">FASTA format
definition</a></p><ul><li>Suffix: <code>.fasta</code> or <code>.fa</code></li><li>Header: none</li><li>Uses: reference genomes for specific organisms, spike-in proteins
and so on</li><li>Acquired from: on-line repositories of reference genomes, Sanger
sequencing</li></ul><p>Rather than a table, a FASTA file is a series of lines of data. This
data can be nucleic acid codes but sometimes amino acid codes. Each
sequence has an identifier and can run to several lines.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="fasta-example">FASTA example<a class="hash-link" href="#fasta-example" title="Direct link to heading">​</a></h3><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">&gt;sequence A</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggtaagtcctctagtacaaacacccccaatattgtgatataattaaaattatattcatat</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">tctgttgccagaaaaaacacttttaggctatattagagccatcttctttgaagcgttgtc</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">&gt;sequence B</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ggtaagtgctctagtacaaacacccccaatattgtgatataattaaaattatattcatat</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">tctgttgccagattttacacttttaggctatattagagccatcttctttgaagcgttgtc</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">tatgcatcgatcgacgactg</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="fastq-format">FASTQ format<a class="hash-link" href="#fastq-format" title="Direct link to heading">​</a></h2><p><a href="https://maq.sourceforge.net/fastq.shtml" target="_blank" rel="noopener noreferrer">FASTQ format definition</a></p><ul><li>Suffix: <code>.fastq</code></li><li>Header: none</li><li>Uses: inputs for mapping and other tools, deposition in on-line
biodata repositories, for a single sample</li><li>Acquired from: sequencing services or machines directly</li></ul><p>A FASTQ file contains a series of lines of data, similar to the FASTA
format. Each sequence fragment has nucleic acid codes, annotated with
quality scores, and an identifier for that fragment. FASTQ files are
usually stored compressed (<code>.fastq.gz</code>).</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="fastq-example">FASTQ example<a class="hash-link" href="#fastq-example" title="Direct link to heading">​</a></h3><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">@EAS54_6_R1_2_1_413_324</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">CCCTTCTTGTCTTCAGCGTTTCTCC</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">;;3;;;;;;;;;;;;7;;;;;;;88</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">@EAS54_6_R1_2_1_540_792</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">TTGGCAGGCCAAGGCCGATGGATCA</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">+</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">;;;;;;;;;;;7;;;;;-;;;3;83</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">@EAS54_6_R1_2_1_443_348</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">GTTGCTTCTGGCGTGGGTGGGGGGG</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">+EAS54_6_R1_2_1_443_348</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">;;;;;;;;;;;9;7;;.7;393333</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="sam-format">SAM format<a class="hash-link" href="#sam-format" title="Direct link to heading">​</a></h2><p><a href="https://samtools.github.io/hts-specs/SAMv1.pdf" target="_blank" rel="noopener noreferrer">SAM format definition</a></p><ul><li>Suffic: <code>.sam</code></li><li>Header: <code>@</code> lines contain extensive information about the data
fields in the main file, the tools and files used to generate them</li><li>Uses: input to variant callers, output from alignment tools,
archival for sequencing projects where space is not an issue</li><li>Acquired from: alignment tools</li></ul><p>This format contains all of the information in a FASTQ file, but also
contains information as to where in the genome the sequence fragments
map to, how well they map there, and how they differ from the reference
for that region.</p><p>It is possible to reconstruct a FASTQ file from a SAM file.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="sam-example">SAM example<a class="hash-link" href="#sam-example" title="Direct link to heading">​</a></h3><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">@HD VN:1.6 SO:coordinate</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">@SQ SN:ref LN:45</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">r001 99 ref 7 30 8M2I4M1D3M = 37 39 TTAGATAAAGGATACTG *</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">r002 0 ref 9 30 3S6M1P1I4M * 0 0 AAAAGATAAGGATA *</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">r003 0 ref 9 30 5S6M * 0 0 GCCTAAGCTAA * SA:Z:ref,29,-,6H5M,17,0;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">r004 0 ref 16 30 6M14N5M * 0 0 ATAGCTTCAGC *</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">r003 2064 ref 29 17 6H5M * 0 0 TAGGC * SA:Z:ref,9,+,5S6M,30,1;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">r001 147 ref 37 30 9M = 7 -39 CAGCGGCAT * NM:i:1</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h3 class="anchor anchorWithStickyNavbar_mojV" id="bam-and-cram-formats">BAM and CRAM formats<a class="hash-link" href="#bam-and-cram-formats" title="Direct link to heading">​</a></h3><p>A <code>.bam</code> file is the binary equivalent of a <code>.sam</code> file, for much more
efficient storage. This is important because <code>.bam</code> files are generally
the largest working files you will encounter (and their <code>.sam</code>
equivalents would often be unworkably so).</p><p>A <code>.cram</code> file is similar, but uses a more finely-tuned compression
algorithm to be even more efficient. It hasn&#x27;t yet achieved the same
level of adoption as the BAM format, and you should check that the tools
you intend to use support it.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="vcf-format">VCF format<a class="hash-link" href="#vcf-format" title="Direct link to heading">​</a></h2><p><a href="https://samtools.github.io/hts-specs/VCFv4.3.pdf" target="_blank" rel="noopener noreferrer">VCF format
definition</a></p><ul><li>Suffix: <code>.vcf</code></li><li>Header: <code>##</code> lines contain information about the sample, how the
sequence was generated, and information about the content of columns
in the file. The <code>#</code> line is the header for the table.</li><li>Uses: showing variants for a sample or comparing variants between
samples</li><li>Acquired from: the output of a variant calling pipeline, downloads
from variant databases</li></ul><p>A format for storing variant information such as single nucleotide
variants, short insertions and deletions, and other sequence variations,
for one or more samples. It only stores differences in a particular
sample from the supplied reference, and hence is smaller than full
sequence data. It has 8 columns containing information about the
variants, then 1 or more columns showing genotypes, 1 column per sample.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="vcf-example">VCF example<a class="hash-link" href="#vcf-example" title="Direct link to heading">​</a></h3><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">##fileformat=VCFv4.3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##fileDate=20090805</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##source=myImputationProgramV3.1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##reference=file:///seq/references/1000GenomesPilot-NCBI36.fasta</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##contig=&lt;ID=20,length=62435964,assembly=B36,md5=f126cdf8a6e0c7f379d618ff66beb2da,species=&quot;Homo sapiens&quot;,taxonomy=x&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##phasing=partial</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##INFO=&lt;ID=NS,Number=1,Type=Integer,Description=&quot;Number of Samples With Data&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##INFO=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Total Depth&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##INFO=&lt;ID=AF,Number=A,Type=Float,Description=&quot;Allele Frequency&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##INFO=&lt;ID=AA,Number=1,Type=String,Description=&quot;Ancestral Allele&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##INFO=&lt;ID=DB,Number=0,Type=Flag,Description=&quot;dbSNP membership, build 129&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##INFO=&lt;ID=H2,Number=0,Type=Flag,Description=&quot;HapMap2 membership&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##FILTER=&lt;ID=q10,Description=&quot;Quality below 10&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##FILTER=&lt;ID=s50,Description=&quot;Less than 50% of samples have data&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=&quot;Genotype&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##FORMAT=&lt;ID=GQ,Number=1,Type=Integer,Description=&quot;Genotype Quality&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##FORMAT=&lt;ID=DP,Number=1,Type=Integer,Description=&quot;Read Depth&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">##FORMAT=&lt;ID=HQ,Number=2,Type=Integer,Description=&quot;Haplotype Quality&quot;&gt;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT NA00001 NA00002 NA00003</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">20 14370 rs6054257 G A 29 PASS NS=3;DP=14;AF=0.5;DB;H2 GT:GQ:DP:HQ 0|0:48:1:51,51 1|0:48:8:51,51 1/1:43:5:.,.</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">20 17330 . T A 3 q10 NS=3;DP=11;AF=0.017 GT:GQ:DP:HQ 0|0:49:3:58,50 0|1:3:5:65,3 0/0:41:3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">20 1110696 rs6040355 A G,T 67 PASS NS=2;DP=10;AF=0.333,0.667;AA=T;DB GT:GQ:DP:HQ 1|2:21:6:23,27 2|1:2:0:18,2 2/2:35:4</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">20 1230237 . T . 47 PASS NS=3;DP=13;AA=T GT:GQ:DP:HQ 0|0:54:7:56,60 0|0:48:4:51,51 0/0:61:2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">20 1234567 microsat1 GTC G,GTCT 50 PASS NS=3;DP=9;AA=G GT:GQ:DP 0/1:35:4 0/2:17:2 1/1:40:3</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h3 class="anchor anchorWithStickyNavbar_mojV" id="bcf-format">BCF format<a class="hash-link" href="#bcf-format" title="Direct link to heading">​</a></h3><p>There is a binary equivalent of VCF, with the same data in a more
compact format.</p><h1>Common tools</h1><p>These formats were usually developed by a team working on a
bioinformatics tool with a need for a consistent input or output format.
There a lot of these formats are closely associated with the original
tool, although most have been adopted as common standards for other
tools working with the same kind of data.</p><p>Unfortunately, although the formats are consistent, the tools themselves
are not with respect to the syntax required to get them to do what you
want. Although they are fully documented, ensuring you are suing them
correctly is not a simple task. This is one reason why it&#x27;s common to
re-use a bioinformatics pipeline that&#x27;s known to work correctly, rather
than trying to create a new, bespoke solution for each project.</p><p>For many tasks there are more than one tool you can use; which you use
should be guided by familiarity with the tool and how easy it is to set
it up and run it. Computing time is cheap. Analyst time is expensive.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="bedtools">Bedtools<a class="hash-link" href="#bedtools" title="Direct link to heading">​</a></h2><p><a href="https://bedtools.readthedocs.io/en/latest/" target="_blank" rel="noopener noreferrer">bedtools documentation</a></p><p>Bedtools is designed to summarise overlaps between genomic regions, or
differences between genomic regions. It was designed around the <code>.bed</code>
format, but can also work with <code>.gff</code>, <code>.bam</code> and <code>.vcf</code> formats.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="samtools">Samtools<a class="hash-link" href="#samtools" title="Direct link to heading">​</a></h2><p><a href="http://www.htslib.org/doc/samtools.html" target="_blank" rel="noopener noreferrer">samtools documentation</a></p><p>Samtools is a Swiss army knife tool for working with <code>.sam</code>, <code>.bam</code>, and
<code>.cram</code> files. It allows you to view regions of the file, convert it to
other formats, manipulate and filter by metadata for reads, produce
summary statistics about sequencing depth and more, and it even has a
primitive genotype caller built-in.</p><p>Many of its functions require its input files to be indexed, and
samtools itself has a mode to index those files.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="bamtools">Bamtools<a class="hash-link" href="#bamtools" title="Direct link to heading">​</a></h2><p><a href="https://hcc.unl.edu/docs/applications/app_specific/bioinformatics_tools/data_manipulation_tools/bamtools/" target="_blank" rel="noopener noreferrer">bamtools
documentation</a></p><p>Bamtools operates solely on <code>.bam</code> files. Whereas samtools includes
features useful for an analyst, bamtools is more concerned with
manipulating the files themselves, to merge, filter or convert.</p><p>Bamtools also has a function to index <code>.bam</code> files.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="vcftools">Vcftools<a class="hash-link" href="#vcftools" title="Direct link to heading">​</a></h2><p><a href="https://vcftools.github.io/index.html" target="_blank" rel="noopener noreferrer">vcftools documentation</a></p><p>Vcftools, predictably, manipulates <code>.vcf</code> files. It can be used to
filter variants, compare variants between files, summarise variants and
handle file conversion, validation and merging.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="bcftools">Bcftools<a class="hash-link" href="#bcftools" title="Direct link to heading">​</a></h2><p><a href="http://www.htslib.org/doc/bcftools.html" target="_blank" rel="noopener noreferrer">bcftools documentation</a></p><p>Bcftools works on both <code>.bcf</code> and <code>.vcf</code> files, and does many of the
same tasks as vcftools.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="bgzip-and-tabix">Bgzip and tabix<a class="hash-link" href="#bgzip-and-tabix" title="Direct link to heading">​</a></h2><p><a href="http://www.htslib.org/doc/bgzip.html" target="_blank" rel="noopener noreferrer">bgzip documentation</a> <a href="http://www.htslib.org/doc/tabix.html" target="_blank" rel="noopener noreferrer">tabix
documentation</a></p><p>These tools are designed to be used together. bgzip is a blocked
implementation of the widely-used gzip compression, and tabix is a tool
for indexing those compressed files.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="igv-integrated-genome-viewer">(IGV) Integrated Genome Viewer<a class="hash-link" href="#igv-integrated-genome-viewer" title="Direct link to heading">​</a></h2><p><a href="https://software.broadinstitute.org/software/igv/" target="_blank" rel="noopener noreferrer">IGV homepage</a></p><p>IGV is a useful tool for visualising data files. It can load a variety
of formats including <code>.bed</code>, <code>.gff</code>, <code>.fasta</code>, <code>.sam</code>, <code>.bam</code> and
<code>.vcf</code>. It&#x27;s invaluable for looking at different files side-by-side,
comparing samples or cross-referencing data and feature definitions.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="igv-screenshot">IGV screenshot<a class="hash-link" href="#igv-screenshot" title="Direct link to heading">​</a></h3><h1>Break</h1><h1>Practical</h1></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/programming/genomic_data/Getting_started_with_genomic_data.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/contributors/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">contributors</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/programming/introduction_to_R/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Introduction to R</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#text-formats" class="table-of-contents__link toc-highlight">Text formats</a></li><li><a href="#binary-formats" class="table-of-contents__link toc-highlight">Binary formats</a></li><li><a href="#compression" class="table-of-contents__link toc-highlight">Compression</a><ul><li><a href="#index-files" class="table-of-contents__link toc-highlight">Index files</a></li></ul></li><li><a href="#file-headers" class="table-of-contents__link toc-highlight">File headers</a></li><li><a href="#versioning" class="table-of-contents__link toc-highlight">Versioning</a></li><li><a href="#bed-format" class="table-of-contents__link toc-highlight">BED format</a><ul><li><a href="#bed-example" class="table-of-contents__link toc-highlight">BED example</a></li></ul></li><li><a href="#gff3-format" class="table-of-contents__link toc-highlight">GFF3 format</a><ul><li><a href="#gff3-example" class="table-of-contents__link toc-highlight">GFF3 example</a></li></ul></li><li><a href="#fasta-format" class="table-of-contents__link toc-highlight">FASTA format</a><ul><li><a href="#fasta-example" class="table-of-contents__link toc-highlight">FASTA example</a></li></ul></li><li><a href="#fastq-format" class="table-of-contents__link toc-highlight">FASTQ format</a><ul><li><a href="#fastq-example" class="table-of-contents__link toc-highlight">FASTQ example</a></li></ul></li><li><a href="#sam-format" class="table-of-contents__link toc-highlight">SAM format</a><ul><li><a href="#sam-example" class="table-of-contents__link toc-highlight">SAM example</a></li><li><a href="#bam-and-cram-formats" class="table-of-contents__link toc-highlight">BAM and CRAM formats</a></li></ul></li><li><a href="#vcf-format" class="table-of-contents__link toc-highlight">VCF format</a><ul><li><a href="#vcf-example" class="table-of-contents__link toc-highlight">VCF example</a></li><li><a href="#bcf-format" class="table-of-contents__link toc-highlight">BCF format</a></li></ul></li><li><a href="#bedtools" class="table-of-contents__link toc-highlight">Bedtools</a></li><li><a href="#samtools" class="table-of-contents__link toc-highlight">Samtools</a></li><li><a href="#bamtools" class="table-of-contents__link toc-highlight">Bamtools</a></li><li><a href="#vcftools" class="table-of-contents__link toc-highlight">Vcftools</a></li><li><a href="#bcftools" class="table-of-contents__link toc-highlight">Bcftools</a></li><li><a href="#bgzip-and-tabix" class="table-of-contents__link toc-highlight">Bgzip and tabix</a></li><li><a href="#igv-integrated-genome-viewer" class="table-of-contents__link toc-highlight">(IGV) Integrated Genome Viewer</a><ul><li><a href="#igv-screenshot" class="table-of-contents__link toc-highlight">IGV screenshot</a></li></ul></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.4b0609aa.js"></script>
<script src="/whg-training-resources/assets/js/main.a59577cb.js"></script>
</body>
</html>