<!doctype html>
<html lang="en" dir="ltr" class="docs-wrapper docs-doc-page docs-version-current plugin-docs plugin-id-default docs-doc-id-programming/introduction_to_R/R_introduction_to_Basic_Features">
<head>
<meta charset="UTF-8">
<meta name="viewport" content="width=device-width,initial-scale=1">
<meta name="generator" content="Docusaurus v2.0.0-beta.20">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.24/dist/katex.min.css" integrity="sha384-odtC+0UGzzFL/6PNoE8rX/SPcQDXBJ+uRepguP4QkPCm2LBxH3FA3y+fKSiJ+AmM" crossorigin="anonymous"><title data-rh="true">R: Introduction to Basic Features | WHG training resources</title><meta data-rh="true" name="twitter:card" content="summary_large_image"><meta data-rh="true" property="og:url" content="https://well.ox.ac.uk//whg-training-resources/programming/introduction_to_R/R_introduction_to_Basic_Features/"><meta data-rh="true" name="docusaurus_locale" content="en"><meta data-rh="true" name="docsearch:language" content="en"><meta data-rh="true" name="docusaurus_version" content="current"><meta data-rh="true" name="docusaurus_tag" content="docs-default-current"><meta data-rh="true" name="docsearch:version" content="current"><meta data-rh="true" name="docsearch:docusaurus_tag" content="docs-default-current"><meta data-rh="true" property="og:title" content="R: Introduction to Basic Features | WHG training resources"><meta data-rh="true" name="description" content="Authors: Helen Lockstone, Matthieu Miossec, Ben Wright Bioinformatics"><meta data-rh="true" property="og:description" content="Authors: Helen Lockstone, Matthieu Miossec, Ben Wright Bioinformatics"><link data-rh="true" rel="icon" href="/whg-training-resources/img/favicon.ico"><link data-rh="true" rel="canonical" href="https://well.ox.ac.uk//whg-training-resources/programming/introduction_to_R/R_introduction_to_Basic_Features/"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/introduction_to_R/R_introduction_to_Basic_Features/" hreflang="en"><link data-rh="true" rel="alternate" href="https://well.ox.ac.uk//whg-training-resources/programming/introduction_to_R/R_introduction_to_Basic_Features/" hreflang="x-default"><link rel="stylesheet" href="/whg-training-resources/assets/css/styles.74b51895.css">
<link rel="preload" href="/whg-training-resources/assets/js/runtime~main.4b0609aa.js" as="script">
<link rel="preload" href="/whg-training-resources/assets/js/main.a59577cb.js" as="script">
</head>
<body class="navigation-with-keyboard">
<script>!function(){function t(t){document.documentElement.setAttribute("data-theme",t)}var e=function(){var t=null;try{t=localStorage.getItem("theme")}catch(t){}return t}();t(null!==e?e:"light")}()</script><div id="__docusaurus">
<div role="region"><a href="#" class="skipToContent_ZgBM">Skip to main content</a></div><nav class="navbar navbar--fixed-top"><div class="navbar__inner"><div class="navbar__items"><button aria-label="Navigation bar toggle" class="navbar__toggle clean-btn" type="button" tabindex="0"><svg width="30" height="30" viewBox="0 0 30 30" aria-hidden="true"><path stroke="currentColor" stroke-linecap="round" stroke-miterlimit="10" stroke-width="2" d="M4 7h22M4 15h22M4 23h22"></path></svg></button><a class="navbar__brand" href="/whg-training-resources/"><div class="navbar__logo"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--light_TfLj"><img src="/whg-training-resources/img/wchg.png" alt="My Site Logo" class="themedImage_W2Cr themedImage--dark_oUvU"></div><b class="navbar__title text--truncate">Training Resources</b></a><a class="navbar__item navbar__link navbar__link--active" href="/whg-training-resources/overview/">Bioinformatics practicals</a></div><div class="navbar__items navbar__items--right"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="navbar__item navbar__link">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a><div class="toggle_S7eR colorModeToggle_vKtC"><button class="clean-btn toggleButton_rCf9 toggleButtonDisabled_Pu9x" type="button" disabled="" title="Switch between dark and light mode (currently light mode)" aria-label="Switch between dark and light mode (currently light mode)"><svg viewBox="0 0 24 24" width="24" height="24" class="lightToggleIcon_v35p"><path fill="currentColor" d="M12,9c1.65,0,3,1.35,3,3s-1.35,3-3,3s-3-1.35-3-3S10.35,9,12,9 M12,7c-2.76,0-5,2.24-5,5s2.24,5,5,5s5-2.24,5-5 S14.76,7,12,7L12,7z M2,13l2,0c0.55,0,1-0.45,1-1s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S1.45,13,2,13z M20,13l2,0c0.55,0,1-0.45,1-1 s-0.45-1-1-1l-2,0c-0.55,0-1,0.45-1,1S19.45,13,20,13z M11,2v2c0,0.55,0.45,1,1,1s1-0.45,1-1V2c0-0.55-0.45-1-1-1S11,1.45,11,2z M11,20v2c0,0.55,0.45,1,1,1s1-0.45,1-1v-2c0-0.55-0.45-1-1-1C11.45,19,11,19.45,11,20z M5.99,4.58c-0.39-0.39-1.03-0.39-1.41,0 c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0s0.39-1.03,0-1.41L5.99,4.58z M18.36,16.95 c-0.39-0.39-1.03-0.39-1.41,0c-0.39,0.39-0.39,1.03,0,1.41l1.06,1.06c0.39,0.39,1.03,0.39,1.41,0c0.39-0.39,0.39-1.03,0-1.41 L18.36,16.95z M19.42,5.99c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06c-0.39,0.39-0.39,1.03,0,1.41 s1.03,0.39,1.41,0L19.42,5.99z M7.05,18.36c0.39-0.39,0.39-1.03,0-1.41c-0.39-0.39-1.03-0.39-1.41,0l-1.06,1.06 c-0.39,0.39-0.39,1.03,0,1.41s1.03,0.39,1.41,0L7.05,18.36z"></path></svg><svg viewBox="0 0 24 24" width="24" height="24" class="darkToggleIcon_nQuB"><path fill="currentColor" d="M9.37,5.51C9.19,6.15,9.1,6.82,9.1,7.5c0,4.08,3.32,7.4,7.4,7.4c0.68,0,1.35-0.09,1.99-0.27C17.45,17.19,14.93,19,12,19 c-3.86,0-7-3.14-7-7C5,9.07,6.81,6.55,9.37,5.51z M12,3c-4.97,0-9,4.03-9,9s4.03,9,9,9s9-4.03,9-9c0-0.46-0.04-0.92-0.1-1.36 c-0.98,1.37-2.58,2.26-4.4,2.26c-2.98,0-5.4-2.42-5.4-5.4c0-1.81,0.89-3.42,2.26-4.4C12.92,3.04,12.46,3,12,3L12,3z"></path></svg></button></div><div class="searchBox_dLyj"></div></div></div><div role="presentation" class="navbar-sidebar__backdrop"></div></nav><div class="main-wrapper docsWrapper_mKqt"><button aria-label="Scroll back to top" class="clean-btn theme-back-to-top-button backToTopButton_RiI4" type="button"></button><div class="docPage_ualW"><aside class="theme-doc-sidebar-container docSidebarContainer_UQUJ"><div class="sidebar_RiAD"><nav class="menu thin-scrollbar menu_izAj"><ul class="theme-doc-sidebar-menu menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/overview/">Overview of the tutorials</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/prerequisites/">Getting your environment setup</a><button aria-label="Toggle the collapsible sidebar category &#x27;Getting your environment setup&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/population_genetics/">Population genetics</a><button aria-label="Toggle the collapsible sidebar category &#x27;Population genetics&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/genome_wide_association_studies/">Genome-wide association analysis</a><button aria-label="Toggle the collapsible sidebar category &#x27;Genome-wide association analysis&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/next_generation_sequencing/">Next-generation sequencing</a><button aria-label="Toggle the collapsible sidebar category &#x27;Next-generation sequencing&#x27;" type="button" class="clean-btn menu__caret"></button></div></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/LICENSE/">LICENSE</a></li><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-1 menu__list-item"><a class="menu__link" href="/whg-training-resources/contributors/">contributors</a></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret menu__link--active" aria-expanded="true" href="/whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/">programming</a></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--sublist-caret" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/genomic_data/Getting_started_with_genomic_data/">genomic_data</a></div></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist menu__link--active" aria-expanded="true" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/">Introduction to R</a><button aria-label="Toggle the collapsible sidebar category &#x27;Introduction to R&#x27;" type="button" class="clean-btn menu__caret"></button></div><ul style="display:block;overflow:visible;height:auto" class="menu__list"><li class="theme-doc-sidebar-item-link theme-doc-sidebar-item-link-level-3 menu__list-item"><a class="menu__link menu__link--active" aria-current="page" tabindex="0" href="/whg-training-resources/programming/introduction_to_R/R_introduction_to_Basic_Features/">R: Introduction to Basic Features</a></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-2 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" tabindex="0" href="/whg-training-resources/programming/programming_with_gene_annotations/">Programming with gene annotations</a><button aria-label="Toggle the collapsible sidebar category &#x27;Programming with gene annotations&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></li><li class="theme-doc-sidebar-item-category theme-doc-sidebar-item-category-level-1 menu__list-item menu__list-item--collapsed"><div class="menu__list-item-collapsible"><a class="menu__link menu__link--sublist" aria-expanded="false" href="/whg-training-resources/statistical_modelling/">Statistical modelling</a><button aria-label="Toggle the collapsible sidebar category &#x27;Statistical modelling&#x27;" type="button" class="clean-btn menu__caret"></button></div></li></ul></nav></div></aside><main class="docMainContainer_uL0j"><div class="container padding-top--md padding-bottom--lg"><div class="row"><div class="col docItemCol_DM6M"><div class="docItemContainer_vinB"><article><nav class="theme-doc-breadcrumbs breadcrumbsContainer_Xlws" aria-label="Breadcrumbs"><ul class="breadcrumbs" itemscope="" itemtype="https://schema.org/BreadcrumbList"><li class="breadcrumbs__item"><a aria-label="Home page" class="breadcrumbs__link" href="/whg-training-resources/"><svg viewBox="0 0 24 24" class="breadcrumbHomeIcon_kU5B"><path d="M10 19v-5h4v5c0 .55.45 1 1 1h3c.55 0 1-.45 1-1v-7h1.7c.46 0 .68-.57.33-.87L12.67 3.6c-.38-.34-.96-.34-1.34 0l-8.36 7.53c-.34.3-.13.87.33.87H5v7c0 .55.45 1 1 1h3c.55 0 1-.45 1-1z" fill="currentColor"></path></svg></a></li><li class="breadcrumbs__item"><span class="breadcrumbs__link">programming</span><meta itemprop="position" content="1"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item"><a class="breadcrumbs__link" itemprop="item" href="/whg-training-resources/programming/introduction_to_R/"><span itemprop="name">Introduction to R</span></a><meta itemprop="position" content="2"></li><li itemscope="" itemprop="itemListElement" itemtype="https://schema.org/ListItem" class="breadcrumbs__item breadcrumbs__item--active"><span class="breadcrumbs__link" itemprop="name">R: Introduction to Basic Features</span><meta itemprop="position" content="3"></li></ul></nav><div class="tocCollapsible_bZGK theme-doc-toc-mobile tocMobile_TmEX"><button type="button" class="clean-btn tocCollapsibleButton_l22C">On this page</button></div><div class="theme-doc-markdown markdown"><h1>R: Introduction to Basic Features</h1><p>Authors: Helen Lockstone, Matthieu Miossec, Ben Wright Bioinformatics
Core</p><h1>Workshop aims</h1><ul><li>To introduce the R language and environment</li><li>To explain some programming jargon and concepts</li><li>To use R to run some typical tasks</li><li>To be aware of the vital importance of good programming habits</li></ul><p>This course is hosted at:
<a href="https://whg-training.github.io/whg-training-resources/programming/introduction_to_R/R_Introduction_to_Basic_Features.html" target="_blank" rel="noopener noreferrer">https://whg-training.github.io/whg-training-resources/programming/introduction_to_R/R_Introduction_to_Basic_Features.html</a>.</p><p>This course has workshop content intended to be worked through using
RStudio. The course tutor will lead work through that content during the
session, but you are encouraged to work at your own pace and read ahead
if you feel comfortable doing so.</p><p>The course is intended to start at the very beginning, for people with
no previous programming experience, using R or otherwise. If you&#x27;re
finding it at too basic a level, feel free to skip ahead or use this
time for something more useful to you! Some suggestions for more R
material are at the end of this document.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="installing-r-and-rstudio">Installing R and RStudio<a class="hash-link" href="#installing-r-and-rstudio" title="Direct link to heading">​</a></h2><p>R and RStudio are available for a variety of platforms, including
Windows, macOS and Linux. R is open source software. R Studio is open
source software, supplemented by non-free add-ons.</p><ul><li><a href="https://cran.r-project.org/bin/windows/base/" target="_blank" rel="noopener noreferrer">R for Windows
download</a></li><li><a href="https://cran.r-project.org/bin/macosx/" target="_blank" rel="noopener noreferrer">R for macOS X</a></li><li>R for Linux can probably be obtained through your usual package
manager, or else <a href="https://cran.r-project.org/bin/linux/" target="_blank" rel="noopener noreferrer">here</a></li><li><a href="https://www.rstudio.com/products/rstudio/download/#download" target="_blank" rel="noopener noreferrer">RStudio
downloads</a></li></ul><p>These should already be installed and ready.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="r-what-is-it">R: what is it?<a class="hash-link" href="#r-what-is-it" title="Direct link to heading">​</a></h2><ul><li>Powerful statistical computing environment and programming language</li><li>Developed by Robert Gentleman and Ross Ihaka at the University of
Auckland in mid 1990s – originally something to teach statistics
that would run on a Mac</li><li><a href="https://rss.onlinelibrary.wiley.com/doi/pdf/10.1111/j.1740-9713.2018.01169.x" target="_blank" rel="noopener noreferrer">Full story of R&#x27;s
development</a></li><li>Now widely used all around the world</li></ul><h2 class="anchor anchorWithStickyNavbar_mojV" id="r-strengths-and-weaknesses">R: strengths and weaknesses<a class="hash-link" href="#r-strengths-and-weaknesses" title="Direct link to heading">​</a></h2><ul><li>Incredibly powerful and versatile statistical programming software…
but where do I start?</li><li>Features that make life easier in many ways… but potential pitfalls
as well</li><li>Open-source, free software with a strong support and development
community</li><li>Extensive additional functionality for bioinformatics and genomic
data through the <a href="https://www.bioconductor.org/" target="_blank" rel="noopener noreferrer">Bioconductor
project</a></li><li>Infamous for opaque error messages and help files</li></ul><h1>Getting started with R</h1><p>This workshop will introduce you to the R environment, with the aim of
becoming familiar with how commands are written and structured and how
to perform basic tasks associated with data analysis using R.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="r-fundamentals">R fundamentals<a class="hash-link" href="#r-fundamentals" title="Direct link to heading">​</a></h2><p>R is, at its heart, a command line language like bash. Text commands are
entered and text output is printed to the console. Similarly, you can
write R scripts to be executed when you need them.</p><p>However, unlike bash, R also keeps data in its own internal store rather
than only keeping data in memory while the command is running. R can
also open new windows to display plots it has created, or show help
files.</p><p>When you close R, it will ask you whether you want to save the
workspace. If you say yes, then R creates a file called &#x27;.Rdata&#x27; (which
is a hidden file on most operating systems by default). This file
contains <em>all</em> of the data you had stored in your R session. This makes
is easy to pick up where you left off, but be aware that this file can
become very large if you load a lot of data into R.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="getting-started-with-rstudio">Getting started with RStudio<a class="hash-link" href="#getting-started-with-rstudio" title="Direct link to heading">​</a></h2><p>RStudio is an integrated development environment (IDE) that wraps around
R and makes a lot of common tasks much more convenient. Launch RStudio
to open a new R session.</p><p>By default, RStudio displays 4 different panels. If there&#x27;s only one
panel on the left, then one of them is currently hidden – you can make
it visible by opening a new R script file from the &#x27;File&#x27; menu.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="rstudio-screenshot">RStudio screenshot<a class="hash-link" href="#rstudio-screenshot" title="Direct link to heading">​</a></h3><p><img loading="lazy" src="/whg-training-resources/assets/images/RStudio_screenshot-cece01842a38154ea4b5f6cd4e4a8150.png" width="1920" height="1080" class="img_E7b_"></p><h3 class="anchor anchorWithStickyNavbar_mojV" id="bottom-left">Bottom left<a class="hash-link" href="#bottom-left" title="Direct link to heading">​</a></h3><p>The bottom left panel is R itself, in the tab &#x27;Console&#x27;. If you were
running R without RStudio, this is exactly what you would see.</p><p>The <code>&gt;</code> prompt is telling you that R is ready to receive new commands.
You can type R code here and press return to have R run that code. R
will print any output from that command and then show the <code>&gt;</code> prompt
again to say it is ready for new input. Note that something may have
happened &#x27;behind the scenes&#x27; – such creating or altering a data object.
It is important to know what each command has done and it was as
intended.</p><p>Two useful things to know about this panel are that you can use the up
and down arrows to scroll through previous commands – very useful to fix
an error without retyping the entire line! – and you can use the tab key
to auto-complete commands, variable names and file names.</p><p>The other tabs you don&#x27;t need to use now.</p><p>Try typing a simple command into the R console (remember to press
Return):</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">6 * 7</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>You&#x27;ll see that it will give you the answer. The <code>[1]</code> is a little
cryptic for now, but that will be explained later. See how it shows the
<code>&gt;</code> prompt again afterwards.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="top-left">Top left<a class="hash-link" href="#top-left" title="Direct link to heading">​</a></h3><p>The top left panel is a simple text editor (like Notepad), intended to
be used to write R scripts. The editor can have multiple files open at a
time.</p><p>This text editor is integrated into RStudio and has several convenient
features. If you click on the green &#x27;Run&#x27; button, it will copy the
current line of code into the R window and run it. If you highlight a
portion of code, the same button will copy and run that selection. There
is also a &#x27;Source&#x27; button, what will copy the entire file into the R
panel and run it.</p><p>Try this by typing this line into the editor:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">6 * 9</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>and clicking &#x27;Run&#x27;.</p><p>The editor also understands some other R-related files. This includes a
&#x27;.rmd&#x27; file. Go to the file menu and select &#x27;Open&#x27;, then navigate to and
select the file for this course: <code>R Introduction to Basic Features.rmd</code>.
You&#x27;ll see this entire document. The snippets of R code each have their
own little run button (the green arrow on the right) that lets you run
the entire section (called a &#x27;chunk&#x27;) with one click.</p><p>Now you have this course open you can follow it from inside RStudio.
Feel free to work through the material at your own pace.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="top-right">Top right<a class="hash-link" href="#top-right" title="Direct link to heading">​</a></h3><p>The top right panel contains information about the current R session.
The &#x27;Environment&#x27; tab lists all the data objects currently stored in R.
Each entry is an &#x27;object&#x27;, in R&#x27;s terminology. The &#x27;History&#x27; tab shows
all the commands that the main R panel has executed so far. This is the
same list you can browse with the arrow keys in the R panel. The other
tabs here you don&#x27;t need to worry about.</p><h3 class="anchor anchorWithStickyNavbar_mojV" id="bottom-right">Bottom right<a class="hash-link" href="#bottom-right" title="Direct link to heading">​</a></h3><p>The bottom right panel is used to show a variety of things in a tabbed
interface. &#x27;Files&#x27; allows you to browse your computer&#x27;s directories.
&#x27;Plots&#x27; displays any plots you have made using R in the current seesion.
&#x27;Help&#x27; is where R will display its help files.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="working-with-the-r-console">Working with the R console<a class="hash-link" href="#working-with-the-r-console" title="Direct link to heading">​</a></h2><p>R is a very interactive environment; commands can be entered into the
console one by one, and each is interpreted and executed by R in real
time. R is a high-level programming language, meaning that it is fairly
human-readable.</p><p>If a command is not valid in the way it is constructed (its syntax), R
will print an error message to the screen. These can sometimes be hard
to interpret but particularly common culprits are simple typing mistakes
or quotes and brackets, whether they are in the wrong place, missing, or
not in pairs (e.g. missing a closing bracket).</p><p>You can type the R commands directly, which is useful at first to get
used to the syntax (structure) of commands. However, you can use
RStudio&#x27;s editor to copy commands into the console if you prefer. This
way, it&#x27;s easier to go back and run a command again, or change it
slightly first.</p><p><em>If you see an error message, first double check that the command is
entered exactly as shown. If you can&#x27;t find the problem, please ask for
help. The advantage of copying the commands into the text editor panel
of RStudio means that you can save the file afterwards to have a record
of what you have done and can easily re-run again in future.</em></p><p>To do anything useful in R, we need to use objects to hold data and
perform various operations on them. The terms object, variable and data
structure can all refer generally to &#x27;objects&#x27; created in R.</p><p>Although variable is a widely used programming term and would be the
preferred term in certain situations, we use &#x27;object&#x27; as a general term
throughout to refer to any of R&#x27;s data structures. These include
vectors, factors, matrices, data frames and lists. Don&#x27;t worry about the
differences between those things for now.</p><p>We will start by entering a few simple commands and discuss what is
happening.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="r-command-syntax">R command syntax<a class="hash-link" href="#r-command-syntax" title="Direct link to heading">​</a></h2><p>The following few commands give some sense of how R stores information
in objects and how simple manipulations of data can be performed. The
<code>#</code> symbol indicates a comment. We&#x27;ll sometimes use them here to show
what R&#x27;s output is without needing to run the command.</p><p>The first example assigns the value 1 to a new object we create and name
<code>x</code>.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- 1</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The assignment operator is <code>&lt;-</code> and running this command creates a new
object named <code>x</code> in R&#x27;s memory. If you look in the &#x27;Environment&#x27; panel
in the top right, you&#x27;ll see that an entry for <code>x</code> has appeared. To
inspect the contents of the new object in the console, we simply type
its name:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">## Inspect the contents of the new object</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] 1</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Note that the contents of an object will be overwritten – without any
warning message – if later assigned a different value:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- 4</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># x now contains the value 4</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] 4</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We can also perform operations directly on the object (note the object
itself does not change):</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x * 2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] 8</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># x still contains the value 4</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] 4</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## unless we were to re-assign the output back to the same object:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- x * 2</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># x now contains the value 8</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] 8</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>R is case sensitive so <code>x</code> and <code>X</code> are different:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">X</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## An error</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Trying to run this command will produce an error message:
<code>Error: object &quot;X&quot; not found</code>. Object names are chosen by the programmer
– informative names are helpful for several reasons. You can use
capitalization, <code>_</code> or <code>.</code> to separate parts of an object name but they
cannot contain spaces, nor start with a number. To avoid confusion or
potential issues, it is also best not to give them the same name as an R
function, which have their own defined names. An object named
<code>raw_data</code>, <code>raw.data</code>, <code>rawData</code> (or even <code>d.raw</code> for minimal typing!)
is fine, but trying to assign a value to a variable named <code>raw data</code>
will give an error because R cannot parse it correctly.</p><p>Elsewhere R ignores whitespace so the following commands are equivalent:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x&lt;-4+3</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- 4 + 3</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Now we will assign a set of numbers to our variable x, which R will
store as a vector. A vector is a series of different data values.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- c(1, 2, 3, 4, 5)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] 1 2 3 4 5</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The <code>[1]</code> part of the output is R&#x27;s way of displaying vectors. The
number in square brackets is the index of the first entry in the line.
If the output is long enough that it spans more than one line, this
makes it easier to work out the index of entries on that line. Note that
in R indices start at <code>1</code> (instead of <code>0</code> as in most other programming
languages).</p><p>If we forget the closing bracket before pressing enter, then R can
detect that the command is incomplete. It gives us a different prompt:
<code>+</code>. This indicates that R is waiting for you to complete the command.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- c(1, 2, 3, 4, 5</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## + appears as the prompt</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If we can&#x27;t simply continue our command, use Esc or control-C to cancel
the input so far and return to the <code>&gt;</code> prompt and start again.</p><p>Rstudio and some other text editors highlight different parts of the
syntax in different colours and automatically close brackets and
quotation marks to help eliminate typing mistakes.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="r-objects">R Objects<a class="hash-link" href="#r-objects" title="Direct link to heading">​</a></h2><p>Objects can be created in many different ways and hold different kinds
of information. Unlike other programming languages, there is no need to
initialise a variable or object in R (i.e. define it before first use) –
it can simply be created and used directly. R also automatically decides
which of its data structures and types are most appropriate for the data
given, rather than being explicitly specified by the programmer.</p><p>We&#x27;ll work through some examples and look at ways to access or
manipulate the data contained within an object. Be aware that the type
(class) of an object and data it contains (numeric, character etc) can
affect how it is treated by R.</p><p>There are many shortcuts in R to avoid tedious or error-prone steps.
When we created our small example vector containing the numbers 1 to 5,
we issued this command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- c(1, 2, 3, 4, 5) # the c function tells R to concatenate these 5 numbers</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## equivalently we could use this command</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- 1:5 # handy if we needed a much longer vector such as 1 to 100, or 1 to 100000</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The <code>:</code> symbol is special syntax that tells R to make a sequence of
whole numbers from one value to another. We can also put together
non-consecutive strings of numbers or a mixture</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x2 &lt;- c(1, 3, 5, 7, 9)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x3 &lt;- c(1:5, 7, 9, 10:15)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>If we need to create a sequence of numbers, the function <code>seq</code> is very
useful.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">seq1 &lt;- seq(from=1, to=99, by=2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">seq2 &lt;- seq(from=0, to=1, by=0.01)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>You can look at all of these objects in the &#x27;Environment&#x27; panel to see
what you get when you use these commands.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="using-functions">Using functions<a class="hash-link" href="#using-functions" title="Direct link to heading">​</a></h2><p>Functions are defined blocks of code that perform a particular task. R
has many hundreds of in-built functions for common tasks, and they can
be used by &#x27;calling&#x27; the function by its name. Above, we used the
function <code>seq</code>; the function name is followed by any arguments in
parentheses – the arguments determine exactly what the function does and
enable it to be useful for many situations. By changing the arguments,
we changed the sequence of numbers that was produced. Some arguments are
required and others are optional or may have default settings defined in
the function. Documentation about the usage of any R function is given
in the corresponding help page:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">help(seq)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em><code>help</code> is, itself, a function that opens up the help file for a given
names function. Note that the help file is opened in the lower right
panel of RStudio.</em></p><h2 class="anchor anchorWithStickyNavbar_mojV" id="vectors">Vectors<a class="hash-link" href="#vectors" title="Direct link to heading">​</a></h2><p>Vectors are one-dimensional objects; in the case of the object <code>x</code> we
created earlier, it has length 5. There is an in-built R function called
length that we can use to check how long any given vector object is:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">length(x)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] 5</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## If we change what is assigned to x, the length of the vector is automatically adjusted:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- 1:10</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">length(x)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] 10</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Vectors are R&#x27;s primary object type and many computations in R are
highly efficient because they operate on the whole vector at once,
rather than element by element.</p><p>Vectors can contain numeric or character data (or both). We can create a
new vector, <code>y</code>, containing the letters &#x27;a&#x27;, &#x27;b&#x27;, and &#x27;c&#x27;.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">y &lt;- c(&quot;a&quot;, &quot;b&quot;, &quot;c&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Note how <code>x</code> and <code>y</code> are displayed differently in the &#x27;Environment&#x27;
panel. We can also see how R has automatically treated them differently
by checking the <code>class</code> of the objects directly:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">class(x)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] &quot;integer&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">class(y)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] &quot;character&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Now try running the following command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">y &lt;- c(a, b, c)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## An error</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>What do you think R has tried to do and why does it result in an error
message?</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="accessing-elements-of-a-vector">Accessing elements of a vector<a class="hash-link" href="#accessing-elements-of-a-vector" title="Direct link to heading">​</a></h2><p>Square brackets are used to access specific elements or subsets of a
vector, factor, matrix or data frame. Let&#x27;s create a new vector as an
example:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x &lt;- c(1:5, 10:14)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## now extract 3rd element</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x[3]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## or extract alternate elements</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x[c(1, 3, 5, 7, 9)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## or extract subset of elements</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x[3:6]</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The second and third examples above show how R can be powerful by
working with vectors. Instead of having to use square brackets to
extract each element of the vector, and then using <code>c</code> to combine them
into a new vector, we use a vector <em>inside</em> the square brackets and R
interprets that as an instruction to return a vector of those elements
of <code>x</code>.</p><p>R will decide the most appropriate way to store the data it is provided
with, and there are ways to convert between different object structures
and classes if needed. To give more examples of how data is interpreted
by R, run the following and note the results (feel free to discuss this
with your neighbours).</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">x2 &lt;- c(1:5, 6.5)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">class(x2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] &quot;numeric&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">x3 &lt;- c(1:5, 6.5, &quot;a&quot;, &quot;b&quot;, &quot;c&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">class(x3)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## [1] &quot;character&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This gives some idea of R&#x27;s internal rules. Because the way data is
being handled by R is important for both performing computations
correctly and the source of many error messages, it is useful to be
familiar with the common data types. Some functions, such as computing a
mean for example, require numeric data objects to operate on. What
happens when you try to find the mean of x2 and x3?</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">mean(x2)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">mean(x3)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="functions-and-packages">Functions and packages<a class="hash-link" href="#functions-and-packages" title="Direct link to heading">​</a></h2><p>There are hundreds, probably thousands, of in-built functions in R. Some
you will use very often and others rarely or never. There are always
several ways to do the same thing in R, using closely-related functions.</p><p>Examples of the functions we have used so far include <code>length</code>, <code>mean</code>,
and <code>class</code>. In the case of <code>length</code>, the argument supplied is the name
of the object we wish to find the length of. We can check the length of
another vector object simply by changing the argument.</p><p>The <code>length</code> function is only applicable to vectors (or factors) and
does not work on other data types such as matrices or data frames, which
are 2-dimensional. Often, data is stored in a table format (e.g. in
Excel), and commonly handled as a &#x27;data frame&#x27; in R. We will introduce
these ideas using practical exercises in the next section.</p><p>There are many thousands of additional functions available to R through
something called &#x27;packages&#x27;. Each package is a self-contained parcel of
functions created for a specific application. For example,
&#x27;Bioconductor&#x27; is a repository of many packages related to handling data
in bioinformatics. Because it would be confusing to always have all
these functions available (and because some packages may end up using
the same names) you need to specifically tell R to make a package
available. This is done using the <code>install.packages</code> and <code>library</code>
functions:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">## Install the package for your R environment</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">install.packages(&quot;ggplot2&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Make the package available to use</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">library(ggplot2)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The first function might prompt you to choose from a list of &#x27;mirrors&#x27;,
which are just different places on the internet where the packages are
stored. It doesn&#x27;t much matter which you choose, although choosing one
close to where you currently are is recommended.</p><p>Note that when installing the package, you need to treat the package
name as character data, inside double-quotes. Once the package has been
installed, R is aware the package exists and can interpret the name
without quotes when you use <code>library</code>. Installing packages is something
you only need to do once per computer (in theory, at least). You must
<code>library</code> in each new R session before trying to use any functions from
the package; as this is what tells R to look in that package for
function names (again, to try and handle duplicate names in different
packages gracefully).</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="working-directory">Working directory<a class="hash-link" href="#working-directory" title="Direct link to heading">​</a></h2><p>An important concept before we get started is the working directory.
Your R session is active in one particular directory on your computer –
by default your &#x27;home&#x27; directory, as defined by the operating system.</p><p>When you save an R session, the <code>.RData</code> file is saved there. When you
start a new R session in a particular directory, any <code>.RData</code> file is
re-opened in that session. If is also possible to load and save these
session files from within RStudio.</p><p>It is useful to set the working directory to the place where you have
the files you want to work with. It means you have to do less typing to
load the save files and you can keep the R version of the data alongside
the original files.</p><p>We need to set a working directory for the rest of the workshop. From
RStudio&#x27;s Session menu, select &#x27;Set working directory&#x27;, and then &#x27;Choose
directory...&#x27;. If you&#x27;ve already downloaded the files for this workshop,
select that directory. Otherwise, create a new directory to work in and
select that. You can change the working directory using the R function
<code>setwd</code>, but it&#x27;s more convenient to use RStudio&#x27;s interface.</p><p>You will also need to download two data files by entering the following
commands:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">download.file(&quot;https://queex.github.io/whg-training-resources/R-intro/inflammation_data.csv&quot;, &quot;./inflammation_data.csv&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">download.file(&quot;https://queex.github.io/whg-training-resources/R-intro/sample.csv&quot;, &quot;./sample.csv&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="removing-objects">Removing objects<a class="hash-link" href="#removing-objects" title="Direct link to heading">​</a></h2><p>Sometimes you are testing things out and creating lots of objects –
that&#x27;s fine but it&#x27;s always good to start a new session when running or
checking your final code to be sure previous objects do not affect it in
any way. Sessions can also be cleaned up by deleting objects with the
<code>rm</code> command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">rm(x3)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h1>Break</h1><p>This is an excellent time to take a short break. Stretch your legs, get
a drink.</p><h1>Workshop part 2</h1><p>Earlier we introduced the R software environment, some key features of
the R programming language and how to start using it. We will now do
some practical exercises working with example data to perform typical
tasks. First, some important acknowledgements:</p><p><em>The tutorial that follows is in part adapted from the <a href="https://software-carpentry.org/lessons/" target="_blank" rel="noopener noreferrer">Software
Carpentry Foundation</a>
Programming with R, specifically the Analysing Patient Data
<a href="http://swcarpentry.github.io/r-novice-inflammation/01-starting-with-data/index.html" target="_blank" rel="noopener noreferrer">tutorial</a>.
The Software Carpentry material is available for re-use under a
<a href="https://creativecommons.org/licenses/by/4.0/" target="_blank" rel="noopener noreferrer">Creative Commons License</a>
and we are grateful to the original authors.</em></p><h2 class="anchor anchorWithStickyNavbar_mojV" id="setup">Setup<a class="hash-link" href="#setup" title="Direct link to heading">​</a></h2><p>We set up our R session for this practical before the break. You can
check the current working directory with the following command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">getwd()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The output (directory path) should match the location you are working in
today, and this folder should also contain the files named
<code>inflammation_data.csv</code> and <code>sample.csv</code>. If you see an error message at
any point, first check the command matches that in the tutorial exactly
and that you haven&#x27;t accidentally missed an earlier command out. Pay
particular attention to lower/upper case letters, underscores, dashes or
dots in function or object names, and that brackets and quotes are
correctly paired. If you can&#x27;t spot the problem or have a question at
any point, please don&#x27;t hesitate to ask!</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="helpful-tips">Helpful Tips<a class="hash-link" href="#helpful-tips" title="Direct link to heading">​</a></h2><ul><li>Enter your commands in the top-left panel of RStudio (a text editor)
as this means they can be saved to keep a record of what you have
done. Use the &#x27;Run&#x27; to check that they run and do what you expect.</li><li>To save your script, click on the disk icon in the same toolbar as
the &#x27;Run&#x27; button. Giving a filename with a .R extension, such as
&#x27;R<!-- -->_<!-- -->course<!-- -->_<!-- -->code.R&#x27;, saves it as an R script file. The <code>.R</code> suffix
lets RStudio know that the file is meant to be an R script.</li><li>Any scripts you make are separate files to the session data that is
saved when you close R.</li></ul><p>This is the usually the easiest way to work; if you need to close your
session and return to it another time, it is easy to run the code again
from start to finish. If it were appropriate, all of the code stored in
a script can be executed in R from start to finish with the command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">source(&quot;script_name.R&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>You may recognise the function name <code>source</code> from the button in the
editor panel that does the same thing.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="reading-in-data-from-a-file">Reading in data from a file<a class="hash-link" href="#reading-in-data-from-a-file" title="Direct link to heading">​</a></h2><p>The first thing we need to do is load or read in the data from our files
so it is accessible in the current R session. There are a few possible
ways to do this but we will use the function <code>read.csv</code> because our
files are saved in &#x27;comma-separated values&#x27; or csv format. To find out
details of how to use this function you can search RStudio&#x27;s Help
(bottom right panel) or type:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">help(read.csv)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The help page shows us the arguments for this function and their default
values where applicable. For example, we see <code>sep=&quot;,&quot;</code>, which means the
fields in a row will be separated on commas, and <code>header=TRUE</code>, which
means it is expecting the first row of the file to contain names for
each column. The first file we will work with is the one named
&#x27;inflammation<!-- -->_<!-- -->data.csv&#x27;. If we inspect this file in Excel or a text
editor (by opening it directly from its location on your computer) we
see there are no column names, just a large set of numerical values.
Therefore, we need to explicitly include the header argument in our
command, changing it to FALSE to over-ride the default behaviour of the
function. This is an example of how arguments modify a function&#x27;s
precise behaviour, rather than requiring two separate functions to exist
for files with/without header rows.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.data &lt;- read.csv(&quot;inflammation_data.csv&quot;, header=FALSE)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>Note: if you see an error message similar to &#x27;No such file or
directory&#x27; when trying to read in a file, it is likely that either (i)
the file is not located in the current working directory, or (ii) there
is one (or more) typos in the filename.</em></p><p>It&#x27;s useful to break this command down a little more.</p><p>We have given a name for a new object, <code>inf.data</code>, in which to store the
contents of the file &#x27;inflammation<!-- -->_<!-- -->data.csv&#x27;. Our object name is
descriptive without being too long, shortening &#x27;inflammation&#x27; to &#x27;inf&#x27;
for our convenience. We are using the in-built R function <code>read.csv</code>,
and provide two arguments:</p><ul><li>the name of the file to read in</li><li><code>header=FALSE</code>, indicating to R that our file does not have a header
row</li></ul><p>There are many other arguments to the <code>read.csv</code> function to further
refine its behaviour but these are either optional or the default
settings are fine for most situations. By running this command, R
creates the object <code>inf.data</code> and information about it appears in the
&#x27;Environment&#x27; panel of RStudio.</p><p>Notice that, unlike the vectors we&#x27;ve created before, that it doesn&#x27;t
show all the data stored in the object. It is detailed as <code>60 obs.</code>
(observations) <code>of 40 variables</code> and if you hover the mouse pointer over
the name, <code>inf.data</code>, it indicates that the object is a <code>data.frame</code>,
which is how R stores tabular data.</p><p>If you double-click on the object name or click on the spreadsheet icon
to the righthand side, the contents of the object are loaded in the
top-left panel in a new tab. The columns have been automatically named
by R as <code>V1</code> through to <code>V40</code>, as column names have to start with a
letter. The rows are simply numbered – the only restriction to row names
is that they must be unique.</p><p>What do you think will happen if you run the command above without
including the header argument? Try it by saving the contents into a new
object called <code>test</code> and compare this to <code>inf.data</code>:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">test &lt;- read.csv(&quot;inflammation_data.csv&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Another object named <code>test</code> now appears. View the contents of that
object and compare it to <code>inf.data</code>.</p><p><em>This is a good example of how easily something unwanted can happen in R
and the importance of checking your objects contain what you intend them
to. Any mistakes can simply be corrected by re-running the command e.g.
with the appropriate header argument, and overwriting any previous
version of the object.</em></p><h2 class="anchor anchorWithStickyNavbar_mojV" id="two-dimensional-data-structures">Two-dimensional data structures<a class="hash-link" href="#two-dimensional-data-structures" title="Direct link to heading">​</a></h2><p>Our original file contained rows and columns of data, and R has suitable
2-dimensional data structures to store such data: matrices and
dataframes. These can both be thought of as tables of data, analogous to
an Excel spreadsheet. Matrices require all columns to be of the same
type, while data frames can have columns of different data types. Given
that experimental data is often a mixture of numeric values (e.g.
measurements) and associated descriptive information, data frames are a
very commonly used data structure in R.</p><p><em>NB while it is possible to create a matrix object with mixed data
types, R will use its internal hierarchy of data types to choose one
that is applicable to all of the data – often this means numerical
columns get converted to character strings, and certain functions may
not perform as expected if this is not noticed, or they may produce an
error message.</em></p><p>In this case, a data frame object has been created:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">class(inf.data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We can check how each column of data has been treated by R, e.g. for the
first column:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">class(inf.data[, 1])</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><em>In fact, as all the columns contain data of the same type, R could
equally well store this data as a matrix object. A data frame has been
created because the functions <code>read.csv</code> and <code>read.table</code> are
specifically designed to deal with mixed column classes and produce data
frames by default. Another function <code>scan</code> can be used to read in
matrices, especially large ones.</em></p><p>It is worth noting that some functions operate on matrix objects, and so
converting between classes is sometimes needed. R has some standard
functions for converting from one class to another.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.data &lt;- as.matrix(inf.data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">class(inf.data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We will continue with the matrix form of this object for now, and load a
mixed dataset later. Most operations on either kind of 2-dimensional
object are the same. For example, we can find the dimensions of a matrix
or a data frame with the <code>dim</code> function.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">dim(inf.data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The output of <code>dim</code> is printed to the screen and shows the number of
rows the object contains, followed by the number of columns (the
convention is always rows, then columns but this can be hard to remember
at first as there is no indication of which way round it is in the
syntax).</p><p>If unsure, the functions <code>nrow</code> and <code>ncol</code> will return the number of
rows or columns respectively:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">nrow(inf.data)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">ncol(inf.data)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This particular file suffers from the lack of any labels to annotate
what data is recorded in the rows and columns. The Software Carpentry
tutorial provides the following information:</p><p><em>“We are studying inflammation in patients who have been given a new
treatment for arthritis. Each row holds the observations for just one
patient. Each column holds the inflammation measured in a day, so we
have a set of values in successive days.&quot;</em></p><p>Our object has 60 rows and 40 columns, so we infer from the information
above that there are 60 patients, and 40 days.</p><p>Again we see a way for mistakes to easily creep into data analysis –
here we have to rely on information given to us second-hand to know what
is what. What if that information were wrong? Are there any checks we
can make ourselves to be sure patients are in rows? We are not told how
many patients were included so simply checking the number of rows won&#x27;t
help. And what if there were 50 patients and measurements taken over 50
days?</p><p>Even with careful scrutiny it would be hard to know how the data are
presented (patients in rows or columns) from the data alone. We could
perhaps make some plots to help us, or we might spot the zero values in
the first column. Scrolling down the object display in the top-left
panel or displaying the first column in the console confirms they are
all zeroes, and the values in each row tend to rise across the first few
columns. We may be reassured by this that the patients are indeed in the
rows, since we might expect inflammation to rise over time, and an
individual recording 0 on every single day might be unlikely (though not
impossible).</p><p>It would be prudent to add some row and column names to reduce the
chance of making a mistake later when dealing with this data:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">rownames(inf.data) &lt;- paste(&quot;Patient&quot;, 1:60, sep=&quot;_&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">colnames(inf.data) &lt;- paste(&quot;Day&quot;, 1:40, sep=&quot;_&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This introduces the very useful and versatile function <code>paste</code>. Note
that adding row and column names does not change the size of the data
object, but we can see them displayed by reloading the object. The row
and column names are not part of the data itself. They are similar to
the alphabetical columns and numbered rows in an Excel spreadsheet.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="accessing-data">Accessing data<a class="hash-link" href="#accessing-data" title="Direct link to heading">​</a></h2><p>Earlier we looked at accessing elements of a one-dimensional vector
object. For matrices and dataframes, a similar approach with square
brackets is used:</p><p><code>object_name[rows,  cols]</code></p><p>By specifying the rows and columns of interest, an object can be subset
in a variety of ways to inspect or extract different parts of it.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[1, 1] # this pulls out the data value in the first row of the first column</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[30, 20] # any single entry can be extracted by specifying the row and column</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Think about how might you select the data in the first 5 rows for the
first 5 columns.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">## If we need to select non-contiguous portions of the object, we&#x27;ll need the help of c():</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[c(1, 3, 5), c(10, 20)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## If you want to display all columns for selected row(s), leave blank space after the comma:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[5, ] # All columns for row 5</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Or blank space before the comma to select all rows for given column(s):</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[, 1:5] # all rows, columns 1 through 5</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>When you select a single column from a data frame, R will &#x27;helpfully&#x27;
turn it into a vector:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">test &lt;- inf.data[, 2] # Select the second column</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">class(test)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This can cause errors if you were expecting <code>test</code> to still be a data
frame in a later command.</p><p>We added column names to our object earlier. Rows and columns can also
be accessed by name (in double-quotes) inside square brackets. <code>$</code> is a
shorthand way of referring to a named column in R. Both alternatives in
the snippet below return the same vector.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.data[, &quot;Day_3&quot;]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">inf.data$Day_3</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Suppose you want to determine the maximum inflammation for patient 5
across days three to seven. To do this you would extract the relevant
subset from the data frame and calculate the maximum value. Which of the
following lines of R code gives the correct answer?</p><ol><li><code>max(inf.data[5, ])</code></li><li><code>max(inf.data[3:7, 5])</code></li><li><code>max(inf.data[5, 3:7])</code></li><li><code>max(inf.data[5, 3, 7])</code></li></ol><h2 class="anchor anchorWithStickyNavbar_mojV" id="analysing-data">Analysing data<a class="hash-link" href="#analysing-data" title="Direct link to heading">​</a></h2><p>We can perform many simple analyses of the data by applying functions
such as max, min, mean, or summary to our data object. We might want to
determine the maximum value per patient or the average value per day.
The following examples illustrate how this can be done extremely
efficiently in R, starting with an approach that is the opposite (and
definitely not recommended!).</p><p>Suppose we want to find the maximum inflammation score for each patient
across the 40 days of measurements. Let&#x27;s start by calculating it for
patient 1.</p><p>Extracting the data for patient 1 (i.e. the first row) is the first
obvious step, and perhaps we decide it makes sense to store the values
for this patient in a new object:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">inf.patient1 &lt;- inf.data[1, ]</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>We can then calculate the maximum value for Patient 1:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">max(inf.patient1)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Although this seems reasonable enough, there are several issues:</p><ul><li>We&#x27;ve created an additional object to store data that is simply a
duplicate of what is already contained in our original object</li><li>It doesn&#x27;t scale well to do this for all 60 patients</li><li>The result is output to the console and therefore hard to do
anything further with</li></ul><p>If we did continue with this approach, there would be 60 new objects
(all with very similar names), a high probability of having made a
typing mistake somewhere (perhaps overwriting one patient&#x27;s data with
another), and a large set of results that we&#x27;d have to manually write
down or transfer to an Excel spreadsheet – all of which is very messy
and prone to error.</p><p>We can easily dispense with the intermediate step of creating a new
object:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain"># max inflammation for patient 1</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">max(inf.data[1, ])</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># or equivalently</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">max(inf.data[&quot;Patient_1&quot;, ])</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>These commands are the same as extracting the data for patient 1 as we
did earlier, but instead of printing to the screen or storing in a new
object, the command is used directly as an argument to the function
<code>max</code> by enclosing in the parentheses.</p><p><em>Commands can be nested in this way to achieve multiple steps in a
single line of code; too many commands in one line though can make it
harder to work out what the code is doing, as well as increase the
chance of the code not doing as intended – the location of brackets
becomes vital.</em></p><p>We&#x27;d really like a way to this for all 60 patients without duplicating
the code 60 times. Loops are one option (not discussed here) but the
<code>apply</code> function is the most efficient approach:</p><p><code>apply</code> allows us to repeat a function on all of the rows (<code>MARGIN = 1</code>)
or columns (<code>MARGIN = 2</code>) of a data frame simultaneously:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">max_inf_patient &lt;- apply(inf.data, MARGIN=1, max)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Similarly, we could compute the average inflammation per day with a
single line of code:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">avg_inf_day &lt;- apply(inf.data, MARGIN=2, mean)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Comparing these two commands will help understand the <code>apply</code> function,
which is not intuitive but highly efficient as we have seen. The
arguments to <code>apply</code> are:</p><ul><li>The data object</li><li><code>MARGIN</code>, indicating whether to apply over rows (1) or columns (2)</li><li>The name of the function to be applied (functions can sometimes be
treated as objects when used as arguments for other functions)</li></ul><p>We wanted to find the maximum inflammation score for each patient, so we
looked across the rows and used the <code>max</code> function. To modify the
command to find the average inflammation per day, we switched the MARGIN
argument to 2 for columns, and gave the final argument as <code>mean</code>.</p><p><em>While the MARGIN argument is explicitly assigned above, R is equally
happy to infer from the shortened command <code>apply(inf.data, 1, max)</code> that
the 1 should be assigned to the second defined argument of <code>apply</code>. You
can also write your own bespoke functions as required and use apply to
run them over an object.</em></p><p>We have also solved the final issue with our initial approach by storing
the results in suitably-named objects for further work.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">length(max_inf_patient)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head(max_inf_patient)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p><code>head</code> is a function that shows the first six elements of a vector or
the first 6 rows of a data frame.</p><p>Another useful function is <code>summary</code>. This returns the minimum value,
first quartile, median, mean, third quartile and the maximum value, all
very useful information to make an initial inspection of your data.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">summary(inf.data[, 1:4]) # for each of the first 4 days</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h2 class="anchor anchorWithStickyNavbar_mojV" id="plotting-data">Plotting data<a class="hash-link" href="#plotting-data" title="Direct link to heading">​</a></h2><p>Visualising data is a vital part of statistical analysis, and R&#x27;s
plotting capabilities are a key reason for its popularity. There is a
related course R: Visualisation that you can take if interested to learn
more. Here, we introduce ways to make a few simple plots. Let&#x27;s take a
look at the average inflammation over time. Recall that we already
calculated these values above and saved them in an object named
<code>avg_inf_day</code>. Plotting the values is done with the function <code>plot</code>:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">plot(avg_inf_day)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Default labels and settings are used but we can refine our plot with some additional arguments:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(avg_inf_day, main=&quot;Inflammation Scores Over Time&quot;, xlab=&quot;Day&quot;, ylab=&quot;Average_inflammation_score&quot;) # adding title and axis labels</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(avg_inf_day, main=&quot;Inflammation Scores Over Time&quot;, xlab=&quot;Day&quot;, ylab=&quot;Average_inflammation_score&quot;, pch=4, col=&quot;red&quot;) # changing the plotting symbols and colour</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Similarly, we could plot the data per patient:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(max_inf_patient)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## Here, we might decide to use a boxplot instead:</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">boxplot(max_inf_patient, main=&quot;Maximum Inflammation Scores&quot;, ylab=&quot;Max_inf_score&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">legend(&quot;topright&quot;, legend=&quot;n=60 patients&quot;, cex=0.8) # adding a legend</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>In RStudio, the &#x27;Plots&#x27; panel in the bottom right contains a history of
all the plots you&#x27;ve made in the current session. In vanilla R, a new
plot will replace the existing one unless you create a new window for
it.</p><p>When we are happy with our plots, they can be saved to a file.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">pdf(&quot;Inflammation_plots.pdf&quot;, onefile=T)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">plot(avg_inf_day, main=&quot;Inflammation Scores Over Time&quot;, xlab=&quot;Day&quot;, ylab=&quot;Average_inflammation_score&quot;, pch=4, col=&quot;red&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">boxplot(max_inf_patient, main=&quot;Maximum Inflammation Scores&quot;, ylab=&quot;Max_inf_score&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">legend(&quot;topright&quot;, legend=&quot;n=60 patients&quot;, cex=0.8)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">dev.off()</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This will be saved to the current working directory by default so if we
check the folder, a new file named &#x27;Inflammation<!-- -->_<!-- -->plots.pdf&#x27; should have
been created. The <code>onefile=T</code> argument instructs R to append additional
plots to the same file and the <code>dev.off()</code> command at the end closes the
file connection. You can also export plots directly to a pdf file from
the RStudio &#x27;Plot&#x27; panel.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="handling-data-with-factors">Handling data with factors<a class="hash-link" href="#handling-data-with-factors" title="Direct link to heading">​</a></h2><p>We&#x27;ll next read in data from another file to illustrate a few more
features of data frames and how to work with them in R. In this case the
file does contain a header row and the default arguments for read.csv
are appropriate for this file so we only need provide the filename:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data2 &lt;- read.csv(&quot;sample.csv&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">head(data2)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This displays the first 6 rows, and we can see immediately that we have
a range of different types of data in each column. Let&#x27;s see how R has
treated it (you can paste the following 4 lines as one block).</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">for(i in 1:ncol(data2)) {</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">    print(class(data2[, i]))</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">}</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Here, we&#x27;ve used a <code>for</code> loop to iterate over each column in the object
<code>data2</code>, and print to screen the class of each column. The output tells
us that columns 1:3 are treated as characters, and mostof the rest are
treated as integers. Column 5 is &#x27;numeric&#x27;, the basic data type in R for
non-integer numbers. R has a number of different classes of data it can
store and sometimes functions will behave differently depending on what
type of data they are given.</p><p>Although columns 1 – 3 are currently characters, some functions in R
that perform statistical analysis prefer a data type called &#x27;factors&#x27;.
Factors are useful when you have category data, where the character data
can only come from a well defined set of possibilities. Factors can be
useful when you want to make sure incorrect values don&#x27;t creep in
accidentally, because R will report an error if factor data tries to be
changed to value outside that set.</p><p>We can specifically convert the Group column to a factor:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data2$Group &lt;- as.factor(data2$Group)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">class(data2$Group)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The <code>as.factor</code> function returns a factor version of a vector. However,
as this doesn&#x27;t change the original <code>data2</code> object by itself, we need to
assign that vector of factors back to the column, replacing what was
there before. You can see how using the <code>$</code> syntax makes this line
easier to read. R has a number of different <code>as.</code> functions for
converting to different data types.</p><p>A very useful summary function is <code>table</code>:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">table(data2$Group)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">table(data2$Gender)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This alerts us to the fact that data in the Gender column has not been
entered consistently, which we might have already spotted from viewing
the object in RStudio.</p><p>To fix this, we can make sure <code>F</code> and <code>M</code> are used throughout; this
involves determining which rows contain a lowercase <code>f</code> for example, and
substituting with <code>F</code>. Similarly for the lowercase case <code>m</code>. The
following code does each a different way.</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">data2$Gender &lt;- gsub(&quot;f&quot;, &quot;F&quot;, data2$Gender)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">data2$Gender[data2$Gender==&quot;m&quot;] &lt;- &quot;M&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain" style="display:inline-block"></span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">## checking we have modified the data as intended</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">table(data2$Gender)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>The <code>gsub</code> function replaces the first argument <code>&quot;f&quot;</code> with the second
argument <code>&quot;F&quot;</code> in the third argument <code>data2$Gender</code> wherever it appears,
leaving the other entries unchanged. As before, we need to assign this
new vector back to the original object.</p><p>The second line shows one of R&#x27;s particularly useful features. We&#x27;re
telling R to replace values in <code>data2$Gender</code>, but only where those
values equal <code>&quot;m&quot;</code>. Indices in R can be used when assigning values to
objects, not just when extracting them.</p><p>This is one example of how there can be multiple ways of doing a
particular task in R. When deciding which method to use, in general, it
is better to stick with the method you understand the best as that&#x27;s the
method where you are least likely to make a mistake. Some mistakes will
cause R to generate an error, but the more dangerous mistakes are ones
that don&#x27;t result in an error but result in incorrect data.</p><h2 class="anchor anchorWithStickyNavbar_mojV" id="putting-it-together">Putting it together<a class="hash-link" href="#putting-it-together" title="Direct link to heading">​</a></h2><p>See if you can solve the following problem using R:</p><ul><li>Create a new vector containing a set of any 9 letters, including at
least 3 vowels. Give your vector object a suitably descriptive name</li><li>Check the length of your vector equals 9</li><li>Now extract subsets of letters to create as many new short words as
you can</li><li>Hint: use the <code>c()</code> function and <code>:</code> operator within square brackets</li></ul><h2 class="anchor anchorWithStickyNavbar_mojV" id="bioconductor">Bioconductor<a class="hash-link" href="#bioconductor" title="Direct link to heading">​</a></h2><p><a href="https://www.bioconductor.org/" target="_blank" rel="noopener noreferrer">Bioconductor</a> is a separate third-party
repository of R code, specifically geared towards bioinformatics. Many
of the packages in Bioconductor are also available via R&#x27;s usual package
installation method. However, Bioconductor has its own preferred
installation mechanism which gets around some of R&#x27;s more annoying
limitations with version incompatibilities. In particular, the packages
in Bioconductor are updated more swiftly following the release of a new
version of R. To get started with Bioconductor, use the following
command:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">source(&quot;https://bioconductor.org/biocLite.R&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>This loads and runs an R script hosted on the web that defines a new
function. You then use this function in a similar way to how you would
use the <code>install.packages</code> function:</p><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">biocLite(&quot;limma&quot;)</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><p>Loading Bioconductor packages this way is typically a longer process but
more reliable.</p><h1>Getting help</h1><p>As you read the R help pages, you may not find them terribly helpful.
The R <code>help</code> function is most useful for refreshing your memory about
specific functions you have used before. Unfortunately, it is not very
useful for learning the how to use the functions in the first place.</p><p>There are several online forums used by R programmers, novice and
expert, to get help and advice from their peers. Searching for your R
problem will often give results from one of these forums.</p><p>One page &#x27;quick reference&#x27; documents:</p><ul><li><a href="https://queex.github.io/whg-training-resources/R-intro/r-cheat-sheet.pdf" target="_blank" rel="noopener noreferrer">R cheat
sheet</a></li><li><a href="https://queex.github.io/whg-training-resources/R-intro/R_reference_card.pdf" target="_blank" rel="noopener noreferrer">R reference
card</a></li></ul><h1>Further resources and useful information</h1><p>The Software Carpentry Foundation website contains many tutorials for
learning a variety of programming languages, including R. There is also
a series of domain-specific Data Carpentry courses, which focus on
computational skills needed to handle and analyse data – tutorials are
currently available for Ecology, Genomics, Geospatial Data and Social
Sciences, with others in development.</p><ul><li><a href="https://software-carpentry.org/lessons/" target="_blank" rel="noopener noreferrer">Software Carpentry
lessons</a></li><li><a href="https://datacarpentry.org/lessons/" target="_blank" rel="noopener noreferrer">Data Carpentry lessons</a></li></ul><p>In particular, the tutorials at the following links give further details
and examples on some of the ideas already introduced or extend to other
topics once you feel comfortable interacting with the R environment.</p><ul><li><a href="http://swcarpentry.github.io/r-novice-inflammation/13-supp-data-structures/index.html" target="_blank" rel="noopener noreferrer">Data
structures</a></li><li><a href="http://swcarpentry.github.io/r-novice-inflammation/11-supp-read-write-csv/index.html" target="_blank" rel="noopener noreferrer">Read and write csv
files</a></li><li><a href="http://swcarpentry.github.io/r-novice-inflammation/12-supp-factors/index.html" target="_blank" rel="noopener noreferrer">Factors</a></li><li><a href="http://swcarpentry.github.io/r-novice-inflammation/10-supp-addressing-data/index.html" target="_blank" rel="noopener noreferrer">Addressing
data</a></li><li><a href="http://swcarpentry.github.io/r-novice-inflammation/06-best-practices-R/index.html" target="_blank" rel="noopener noreferrer">Best
practices</a></li><li><a href="https://datacarpentry.org/R-genomics/" target="_blank" rel="noopener noreferrer">R genomics</a></li><li><a href="https://datacarpentry.org/r-intro-geospatial/" target="_blank" rel="noopener noreferrer">Geospatial data</a></li></ul><h1>Putting it together example solution</h1><div class="codeBlockContainer_MPoW theme-code-block" style="--prism-color:#393A34;--prism-background-color:#f6f8fa"><div class="codeBlockContent_B9tL"><pre tabindex="0" class="prism-code language-text codeBlock__0OG thin-scrollbar"><code class="codeBlockLines_gEuF"><span class="token-line" style="color:#393A34"><span class="token plain">random_letters &lt;- c(&quot;r&quot;, &quot;t&quot;, &quot;d&quot;, &quot;e&quot;, &quot;e&quot;, &quot;i&quot;, &quot;a&quot;, &quot;s&quot;, &quot;n&quot;)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">length(random_letters)</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># [1] 9</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">random_letters[c(8, 7, 6, 3)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># [1] &quot;s&quot; &quot;a&quot; &quot;i&quot; &quot;d&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">random_letters[c(1, 7, 6, 8, 4)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># [1] &quot;r&quot; &quot;a&quot; &quot;i&quot; &quot;s&quot; &quot;e&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">random_letters[c(8, 2, 4:5, 1)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># [1] &quot;s&quot; &quot;t&quot; &quot;e&quot; &quot;e&quot; &quot;r&quot;</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain">random_letters[c(2, 1, 4, 4)]</span><br></span><span class="token-line" style="color:#393A34"><span class="token plain"># [1] &quot;t&quot; &quot;r&quot; &quot;e&quot; &quot;e&quot;</span><br></span></code></pre><div class="buttonGroup_hRr1"><button type="button" aria-label="Copy code to clipboard" title="Copy" class="clean-btn"><span class="copyButtonIcons_W9eQ" aria-hidden="true"><svg class="copyButtonIcon_XEyF" viewBox="0 0 24 24"><path d="M19,21H8V7H19M19,5H8A2,2 0 0,0 6,7V21A2,2 0 0,0 8,23H19A2,2 0 0,0 21,21V7A2,2 0 0,0 19,5M16,1H4A2,2 0 0,0 2,3V17H4V3H16V1Z"></path></svg><svg class="copyButtonSuccessIcon_i9w9" viewBox="0 0 24 24"><path d="M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z"></path></svg></span></button></div></div></div><h1>Epilogue: good programming habits</h1><ol><li>Break it down into small steps<ul><li>Write code to do individual parts of your overall task and build
it up into a larger script</li><li>This makes it easier to get right and test as you are going
along</li><li>It is also a good way to solve errors/bugs in your code</li></ul></li><li>Pay attention to details and program carefully<ul><li>Commands must be accurate both in their syntax and what they do</li><li>It is easy to overwrite objects, pull out wrong subsets of data,
mix up sample names etc.</li><li>To avoid this, you need to check carefully all the time what you
are working with and what R produces after you run a command</li><li>Inspect object contents, make plots, test examples can all
helpGood programming habits</li></ul></li><li>Test your code thoroughly<ul><li>You are the only person who can make sure it is doing what you
intend it to do</li><li>R will give an error if a command is incorrectly specified
according to the rules of the language but it can&#x27;t tell you if
the command you&#x27;ve written does not do what you think it does…</li><li>…and it will go ahead and execute it regardless, because that is
what you have instructed R to do!</li></ul></li><li>Comment your code<ul><li>Comments are really useful to jog your memory if you come back
to some code after a while or want someone else to understand
it.</li></ul></li></ol><h1>Finally…</h1><p>Please let us know if you have any questions and good luck on your
programming journey!</p><p>We are happy to help anytime:
<a href="mailto://bioinformatics@well.ox.ac.uk" target="_blank" rel="noopener noreferrer">bioinformatics@well.ox.ac.uk</a></p></div><footer class="theme-doc-footer docusaurus-mt-lg"><div class="theme-doc-footer-edit-meta-row row"><div class="col"><a href="https://github.com/whg-training/whg-training-resources/docs/programming/introduction_to_R/R_introduction_to_Basic_Features.md" target="_blank" rel="noreferrer noopener" class="theme-edit-this-page"><svg fill="currentColor" height="20" width="20" viewBox="0 0 40 40" class="iconEdit_dcUD" aria-hidden="true"><g><path d="m34.5 11.7l-3 3.1-6.3-6.3 3.1-3q0.5-0.5 1.2-0.5t1.1 0.5l3.9 3.9q0.5 0.4 0.5 1.1t-0.5 1.2z m-29.5 17.1l18.4-18.5 6.3 6.3-18.4 18.4h-6.3v-6.2z"></path></g></svg>Edit this page</a></div><div class="col lastUpdated_foO9"></div></div></footer></article><nav class="pagination-nav docusaurus-mt-lg" aria-label="Docs pages navigation"><a class="pagination-nav__link pagination-nav__link--prev" href="/whg-training-resources/programming/introduction_to_R/"><div class="pagination-nav__sublabel">Previous</div><div class="pagination-nav__label">Introduction to R</div></a><a class="pagination-nav__link pagination-nav__link--next" href="/whg-training-resources/programming/programming_with_gene_annotations/"><div class="pagination-nav__sublabel">Next</div><div class="pagination-nav__label">Programming with gene annotations</div></a></nav></div></div><div class="col col--3"><div class="tableOfContents_cNA8 thin-scrollbar theme-doc-toc-desktop"><ul class="table-of-contents table-of-contents__left-border"><li><a href="#installing-r-and-rstudio" class="table-of-contents__link toc-highlight">Installing R and RStudio</a></li><li><a href="#r-what-is-it" class="table-of-contents__link toc-highlight">R: what is it?</a></li><li><a href="#r-strengths-and-weaknesses" class="table-of-contents__link toc-highlight">R: strengths and weaknesses</a></li><li><a href="#r-fundamentals" class="table-of-contents__link toc-highlight">R fundamentals</a></li><li><a href="#getting-started-with-rstudio" class="table-of-contents__link toc-highlight">Getting started with RStudio</a><ul><li><a href="#rstudio-screenshot" class="table-of-contents__link toc-highlight">RStudio screenshot</a></li><li><a href="#bottom-left" class="table-of-contents__link toc-highlight">Bottom left</a></li><li><a href="#top-left" class="table-of-contents__link toc-highlight">Top left</a></li><li><a href="#top-right" class="table-of-contents__link toc-highlight">Top right</a></li><li><a href="#bottom-right" class="table-of-contents__link toc-highlight">Bottom right</a></li></ul></li><li><a href="#working-with-the-r-console" class="table-of-contents__link toc-highlight">Working with the R console</a></li><li><a href="#r-command-syntax" class="table-of-contents__link toc-highlight">R command syntax</a></li><li><a href="#r-objects" class="table-of-contents__link toc-highlight">R Objects</a></li><li><a href="#using-functions" class="table-of-contents__link toc-highlight">Using functions</a></li><li><a href="#vectors" class="table-of-contents__link toc-highlight">Vectors</a></li><li><a href="#accessing-elements-of-a-vector" class="table-of-contents__link toc-highlight">Accessing elements of a vector</a></li><li><a href="#functions-and-packages" class="table-of-contents__link toc-highlight">Functions and packages</a></li><li><a href="#working-directory" class="table-of-contents__link toc-highlight">Working directory</a></li><li><a href="#removing-objects" class="table-of-contents__link toc-highlight">Removing objects</a></li><li><a href="#setup" class="table-of-contents__link toc-highlight">Setup</a></li><li><a href="#helpful-tips" class="table-of-contents__link toc-highlight">Helpful Tips</a></li><li><a href="#reading-in-data-from-a-file" class="table-of-contents__link toc-highlight">Reading in data from a file</a></li><li><a href="#two-dimensional-data-structures" class="table-of-contents__link toc-highlight">Two-dimensional data structures</a></li><li><a href="#accessing-data" class="table-of-contents__link toc-highlight">Accessing data</a></li><li><a href="#analysing-data" class="table-of-contents__link toc-highlight">Analysing data</a></li><li><a href="#plotting-data" class="table-of-contents__link toc-highlight">Plotting data</a></li><li><a href="#handling-data-with-factors" class="table-of-contents__link toc-highlight">Handling data with factors</a></li><li><a href="#putting-it-together" class="table-of-contents__link toc-highlight">Putting it together</a></li><li><a href="#bioconductor" class="table-of-contents__link toc-highlight">Bioconductor</a></li></ul></div></div></div></div></main></div></div><footer class="footer footer--dark"><div class="container container-fluid"><div class="row footer__links"><div class="col footer__col"><div class="footer__title">License</div><ul class="footer__items clean-list"><li class="footer__item"><a class="footer__link-item" href="/whg-training-resources/LICENSE/">License</a></li></ul></div><div class="col footer__col"><div class="footer__title">More</div><ul class="footer__items clean-list"><li class="footer__item"><a href="https://github.com/whg-training/whg-training-resources" target="_blank" rel="noopener noreferrer" class="footer__link-item">GitHub<svg width="13.5" height="13.5" aria-hidden="true" viewBox="0 0 24 24" class="iconExternalLink_I5OW"><path fill="currentColor" d="M21 13v10h-21v-19h12v2h-10v15h17v-8h2zm3-12h-10.988l4.035 4-6.977 7.07 2.828 2.828 6.977-7.07 4.125 4.172v-11z"></path></svg></a></li></ul></div></div><div class="footer__bottom text--center"><div class="footer__copyright">Copyright © 2022 University of Oxford.  Built with Docusaurus.</div></div></div></footer></div>
<script src="/whg-training-resources/assets/js/runtime~main.4b0609aa.js"></script>
<script src="/whg-training-resources/assets/js/main.a59577cb.js"></script>
</body>
</html>